### ISSUE TITLE
audacity crashes on resize (fit to height)

### ISSUE THREAD

**Issue by dvzrv** [audacity/audacity#538]
**Describe the bug**
It seems that audacity 2.4.1 on Arch Linux suffers from a bug in regards to resizing to "fit to height" ([upstream report](https://bugs.archlinux.org/task/66771)).

**To Reproduce**
Steps to reproduce:
* Create a track to work with (Generate > Tone, for example) or open an audio file.
* Select Track Size > Fit to Height from the View menu.
* App crashes without displaying the usual crash dialog.

**Expected behavior**
Audacity does not crash

**Screenshots**

**Additional information (please complete the following information):**
 - OS: Arch Linux
 - Version Audacity 2.4.1

**Additional context**
According to the forum post linked to in the downstream ticket @SteveDaulton mentions, that this is [a "known problem" due to autotools being used instead of cmake](https://forum.audacityteam.org/viewtopic.php?p=395731#p395731).
While I generally welcome the switch to cmake, I find the above statement rather interesting as I can not find any information about this *known issue* on this issue tracker, [on the old issue tracker](https://bugzilla.audacityteam.org/buglist.cgi?bug_status=__all__&content=resize&no_redirect=1&order=Importance&product=&query_format=specific) or in the release notes of 2.4.0 or 2.4.1.

Given, that the cmake setup enforces the use of python2 (#520) (which the autotools setup does not) and has numerous other unresolved issues (#519, #521, #522) the new setup is currently not really a viable alternative.

From a packaging perspective it would have been nice to have more transparency about the new cmake setup and a longer grace period in regards to it as this is a lot of additional work, that could have been avoided.

**Comment by dvzrv** [comment]
Here is a stacktrace using coredumpctl of this crash:

```
           PID: 1747547 (audacity)
           UID: 1000 (dave)
           GID: 1000 (dave)
        Signal: 11 (SEGV)
     Timestamp: Mon 2020-05-25 19:43:58 CEST (2min 50s ago)
  Command Line: audacity
    Executable: /usr/bin/audacity
 Control Group: /user.slice/user-1000.slice/session-2.scope
          Unit: session-2.scope
         Slice: user-1000.slice
       Session: 2
     Owner UID: 1000 (dave)
       Boot ID: 865656356fcb4728b4901fea125a8be3
    Machine ID: c14dea973584408b98c4c1f443221ed8
      Hostname: dvzrv
       Storage: /var/lib/systemd/coredump/core.audacity.1000.865656356fcb4728b4901fea125a8be3.1747547.1590428638000000000000.lz4
       Message: Process 1747547 (audacity) of user 1000 dumped core.
                
                Stack trace of thread 1747547:
                #0  0x0000562fa2f20e27 _ZN11ViewActions7Handler10OnZoomFitVERK14CommandContext (audacity + 0xa3ce27)
                #1  0x0000562fa2d26c85 _ZN14CommandManager18HandleCommandEntryER15AudacityProjectPK16CommandListEntrySt6bitsetILm64EEbPK7wxEvent (audacity + 0x842c85)
                #2  0x0000562fa2d2bac6 _ZN14CommandManager12HandleMenuIDER15AudacityProjectiSt6bitsetILm64EEb (audacity + 0x847ac6)
                #3  0x0000562fa2c619f0 _ZN13ProjectWindow6OnMenuER14wxCommandEvent (audacity + 0x77d9f0)
                #4  0x00007f00f9ad94c1 _ZN12wxEvtHandler23ProcessEventIfMatchesIdERK21wxEventTableEntryBasePS_R7wxEvent (libwx_baseu-3.0.so.0 + 0x20a4c1)
                #5  0x00007f00f9ad95c4 _ZN16wxEventHashTable11HandleEventER7wxEventP12wxEvtHandler (libwx_baseu-3.0.so.0 + 0x20a5c4)
                #6  0x00007f00f9ad990e _ZN12wxEvtHandler11TryHereOnlyER7wxEvent (libwx_baseu-3.0.so.0 + 0x20a90e)
                #7  0x00007f00f9ad9783 _ZN12wxEvtHandler10DoTryChainER7wxEvent (libwx_baseu-3.0.so.0 + 0x20a783)
                #8  0x00007f00f9ad9a32 _ZN12wxEvtHandler12ProcessEventER7wxEvent (libwx_baseu-3.0.so.0 + 0x20aa32)
                #9  0x00007f00fa03df4a _ZN12wxWindowBase8TryAfterER7wxEvent (libwx_gtk3u_core-3.0.so.0 + 0x46af4a)
                #10 0x00007f00f9ad97c7 _ZN12wxEvtHandler18SafelyProcessEventER7wxEvent (libwx_baseu-3.0.so.0 + 0x20a7c7)
                #11 0x00007f00f9ff028f _ZN10wxMenuBase9SendEventEii (libwx_gtk3u_core-3.0.so.0 + 0x41d28f)
                #12 0x00007f00f9edb2e0 n/a (libwx_gtk3u_core-3.0.so.0 + 0x3082e0)
                #13 0x00007f00f8d93052 g_closure_invoke (libgobject-2.0.so.0 + 0x13052)
                #14 0x00007f00f8dbb685 n/a (libgobject-2.0.so.0 + 0x3b685)
                #15 0x00007f00f8daf41d g_signal_emit_valist (libgobject-2.0.so.0 + 0x2f41d)
                #16 0x00007f00f8daf980 g_signal_emit (libgobject-2.0.so.0 + 0x2f980)
                #17 0x00007f00f90370db gtk_widget_activate (libgtk-3.so.0 + 0x13f0db)
                #18 0x00007f00f916c95d gtk_menu_shell_activate_item (libgtk-3.so.0 + 0x27495d)
                #19 0x00007f00f916f494 n/a (libgtk-3.so.0 + 0x277494)
                #20 0x00007f00f92cfe1c n/a (libgtk-3.so.0 + 0x3d7e1c)
                #21 0x00007f00f8daf80a g_signal_emit_valist (libgobject-2.0.so.0 + 0x2f80a)
                #22 0x00007f00f8daf980 g_signal_emit (libgobject-2.0.so.0 + 0x2f980)
                #23 0x00007f00f9038c26 n/a (libgtk-3.so.0 + 0x140c26)
                #24 0x00007f00f917f095 n/a (libgtk-3.so.0 + 0x287095)
                #25 0x00007f00f918262c gtk_main_do_event (libgtk-3.so.0 + 0x28a62c)
                #26 0x00007f00f8e93234 n/a (libgdk-3.so.0 + 0x95234)
                #27 0x00007f00f8e626f4 n/a (libgdk-3.so.0 + 0x646f4)
                #28 0x00007f00f8ca943c g_main_context_dispatch (libglib-2.0.so.0 + 0x5243c)
                #29 0x00007f00f8cf6fa9 n/a (libglib-2.0.so.0 + 0x9ffa9)
                #30 0x00007f00f8ca7c03 g_main_loop_run (libglib-2.0.so.0 + 0x50c03)
                #31 0x00007f00f917bc0f gtk_main (libgtk-3.so.0 + 0x283c0f)
                #32 0x00007f00f9e5a856 _ZN14wxGUIEventLoop5DoRunEv (libwx_gtk3u_core-3.0.so.0 + 0x287856)
                #33 0x00007f00f998fe0e _ZN15wxEventLoopBase3RunEv (libwx_baseu-3.0.so.0 + 0xc0e0e)
                #34 0x00007f00f9954f27 _ZN16wxAppConsoleBase8MainLoopEv (libwx_baseu-3.0.so.0 + 0x85f27)
                #35 0x00007f00f99e37f9 _Z7wxEntryRiPPw (libwx_baseu-3.0.so.0 + 0x1147f9)
                #36 0x0000562fa2a25b93 main (audacity + 0x541b93)
                #37 0x00007f00f862e002 __libc_start_main (libc.so.6 + 0x27002)
                #38 0x0000562fa2aa01ee _start (audacity + 0x5bc1ee)
                
                Stack trace of thread 1747766:
                #0  0x00007f00f8b19e32 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0 + 0xfe32)
                #1  0x00007f00f8b595f0 n/a (libjack.so.0 + 0x2d5f0)
                #2  0x00007f00f8b50e16 n/a (libjack.so.0 + 0x24e16)
                #3  0x00007f00f8b58b1d n/a (libjack.so.0 + 0x2cb1d)
                #4  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #5  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747550:
                #0  0x00007f00f86fc05f __poll (libc.so.6 + 0xf505f)
                #1  0x00007f00f8cf6f38 n/a (libglib-2.0.so.0 + 0x9ff38)
                #2  0x00007f00f8ca7c03 g_main_loop_run (libglib-2.0.so.0 + 0x50c03)
                #3  0x00007f00f80551a8 n/a (libgio-2.0.so.0 + 0x1001a8)
                #4  0x00007f00f8cd1511 n/a (libglib-2.0.so.0 + 0x7a511)
                #5  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #6  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747768:
                #0  0x00007f00f870171d syscall (libc.so.6 + 0xfa71d)
                #1  0x00007f00f8b5a93b n/a (libjack.so.0 + 0x2e93b)
                #2  0x00007f00f8b410c5 n/a (libjack.so.0 + 0x150c5)
                #3  0x00007f00f8b401f7 n/a (libjack.so.0 + 0x141f7)
                #4  0x00007f00f8b3fa08 n/a (libjack.so.0 + 0x13a08)
                #5  0x00007f00f8b58b1d n/a (libjack.so.0 + 0x2cb1d)
                #6  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #7  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747548:
                #0  0x00007f00f86fc05f __poll (libc.so.6 + 0xf505f)
                #1  0x00007f00f8cf6f38 n/a (libglib-2.0.so.0 + 0x9ff38)
                #2  0x00007f00f8ca8221 g_main_context_iteration (libglib-2.0.so.0 + 0x51221)
                #3  0x00007f00f8ca8272 n/a (libglib-2.0.so.0 + 0x51272)
                #4  0x00007f00f8cd1511 n/a (libglib-2.0.so.0 + 0x7a511)
                #5  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #6  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747769:
                #0  0x00007f00f86ce571 clock_nanosleep@@GLIBC_2.17 (libc.so.6 + 0xc7571)
                #1  0x00007f00f86d3e87 __nanosleep (libc.so.6 + 0xcce87)
                #2  0x00007f00f9aac3cd _Z12wxMicroSleepm (libwx_baseu-3.0.so.0 + 0x1dd3cd)
                #3  0x0000562fa2b10d92 _ZN10MidiThread5EntryEv (audacity + 0x62cd92)
                #4  0x00007f00f9aa7e3a _ZN8wxThread9CallEntryEv (libwx_baseu-3.0.so.0 + 0x1d8e3a)
                #5  0x00007f00f9aa8261 n/a (libwx_baseu-3.0.so.0 + 0x1d9261)
                #6  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #7  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747767:
                #0  0x00007f00f8b1d87c read (libpthread.so.0 + 0x1387c)
                #1  0x00007f00f8b59d6f n/a (libjack.so.0 + 0x2dd6f)
                #2  0x00007f00f8b5ec59 n/a (libjack.so.0 + 0x32c59)
                #3  0x00007f00f8b58b1d n/a (libjack.so.0 + 0x2cb1d)
                #4  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #5  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747551:
                #0  0x00007f00f86fc05f __poll (libc.so.6 + 0xf505f)
                #1  0x00007f00f8cf6f38 n/a (libglib-2.0.so.0 + 0x9ff38)
                #2  0x00007f00f8ca8221 g_main_context_iteration (libglib-2.0.so.0 + 0x51221)
                #3  0x00007f00fa67be5e n/a (libdconfsettings.so + 0x4e5e)
                #4  0x00007f00f8cd1511 n/a (libglib-2.0.so.0 + 0x7a511)
                #5  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #6  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
                
                Stack trace of thread 1747770:
                #0  0x00007f00f86ce571 clock_nanosleep@@GLIBC_2.17 (libc.so.6 + 0xc7571)
                #1  0x00007f00f86d3e87 __nanosleep (libc.so.6 + 0xcce87)
                #2  0x00007f00f9aac3cd _Z12wxMicroSleepm (libwx_baseu-3.0.so.0 + 0x1dd3cd)
                #3  0x0000562fa2b15500 _ZN11AudioThread5EntryEv (audacity + 0x631500)
                #4  0x00007f00f9aa7e3a _ZN8wxThread9CallEntryEv (libwx_baseu-3.0.so.0 + 0x1d8e3a)
                #5  0x00007f00f9aa8261 n/a (libwx_baseu-3.0.so.0 + 0x1d9261)
                #6  0x00007f00f8b13422 start_thread (libpthread.so.0 + 0x9422)
                #7  0x00007f00f8706bf3 __clone (libc.so.6 + 0xffbf3)
```

**Comment by SteveDaulton** [comment]
The "Track Size > Fit to Height -> Crash" issue was discussed on the developer's mailing list.
I did not log it on the Audacity bug tracker because I was told by the release manager that autotools build chain is no longer supported, and I could not reproduce the issue with a cmake build.

Personally I think the Linux build.txt could be clearer, an no doubt will be in future versions, but it does already say to use cmake for building Audacity rather than autotools.

The issue of the Python 2 requirement is logged on the Audacity bug tracker:
https://bugzilla.audacityteam.org/show_bug.cgi?id=2440
https://bugzilla.audacityteam.org/show_bug.cgi?id=2441

The move from autotools to cmake was discussed on multiple occasions on the public Audacity developers mailing list, for example, back in February: https://sourceforge.net/p/audacity/mailman/message/36921276/

I assume that you are not subscribed to that list, so will not have been aware of the change. Where could we have mentioned the change from autotools to cmake where you would have seen it?

**Comment by dvzrv** [comment]
> I assume that you are not subscribed to that list, so will not have been aware of the change. Where could we have mentioned the change from autotools to cmake where you would have seen it?

No, I am on many many mailing lists though and this would probably fly right by me anyhow (depends on the traffic though ;-) ).
A good place to document this is always the changelog (and just using the same information in the release information here on github or wherever you want to announce it). That being said, audacity has a `CHANGELOG.txt`, but it seems its use has ceased with 2.3.3.

Ideally, one would do something of the sort to switch build systems:

* Version a.b.c: With this version Audacity introduces cmake as an alternative build system
* Version a.b.d: With this version Audacity has established cmake as a full alternative to the previously default autotools based build system
* Version a.c.a: With this version Audacity makes cmake the default build system. The autotools build system is in maintenance mode only.
* Version a.d.a: With this version Audacity deprecates and removes the autotools build system.

Anything else is kind of unfair towards downstream as the current situation is now rather an "all or nothing" type of situation.

**Comment by dvzrv** [comment]
@SteveDaulton also, thanks for the clarification!

**Comment by JamesCrook** [comment]
It is actually logged in our Bugzilla.  Steve did log it.
https://bugzilla.audacityteam.org/show_bug.cgi?id=2421

I reduced it from P1 (blocks our release) to P2.  

Patches for CMake are welcome, provided they do not break our builds.  (Windows 10, Mac 10.15, Ubuntu 18).  We too would like to be free of Python 2 dependency.  We are currently working to release a 2.4.2 that shifts us to wx3.1.3 (https://wiki.audacityteam.org/wiki/Next_Release)

**Comment by Barracuda72** [comment]
> It is actually logged in our Bugzilla. Steve did log it.

The problem is, for some reason Audacitys' entire site is tainted by Cloudfare. That's an absolute no-go for a free software project, IMHO.

Anyway, back to the issue. Is there any clues about what is causing this weird behaviour? Why it only happens in autotools build? Maybe it's not an autotools issue, but it just so happens to reveal itself there?

**Comment by lllucius** [comment]
> It is actually logged in our Bugzilla. Steve did log it.
> The problem is, for some reason Audacitys' entire site is tainted by Cloudfare.
> That's an absolute no-go for a free software project, IMHO.

You mean that following good security practices to protect our sites from malicious actors goes against the tenants of free software?  I'd absolutely love to read that article...please share a link.

> Anyway, back to the issue. Is there any clues about what is causing this weird
> behaviour? Why it only happens in autotools build? Maybe it's not an
> autotools issue, but it just so happens to reveal itself there?

Prior to our removal of the autotools build, I'd tracked down the crash to somewhere between commits a465ce0046e1726b3d2d926091b91714e154a91e (no crash) and c1d129ec9863a877cdd678695fff54a5dab5a191 (crash).  I never did figure out why it happened with autotools and not with cmake.

**Comment by Barracuda72** [comment]
> You mean that following good security practices to protect our sites from malicious actors goes against the tenants of free software? I'd absolutely love to read that article...please share a link.

Your mean, injecting proven to be malicious third-party service, compromising your website by your own hands and depriving users of their freedom is now called "good security practices"? WOW, that's really a long stretch...

If you want some links, you can just search for something like "Why Cloudfare is Bad". Here's only a couple:
https://lists.gnu.org/archive/html/gnu-system-discuss/2017-03/msg00126.html
http://techrights.org/2019/02/17/the-cloudflare-trap/
https://pixie.garden/~/ThufiesBlog/don%27t-trust-cloud-flare (beware, certificate expired, but site itself isn't dubious and the article is very worth reading)

> Prior to our removal of the autotools build, I'd tracked down the crash to somewhere between commits [a465ce0](https://github.com/audacity/audacity/commit/a465ce0046e1726b3d2d926091b91714e154a91e) (no crash) and [c1d129e](https://github.com/audacity/audacity/commit/c1d129ec9863a877cdd678695fff54a5dab5a191) (crash). I never did figure out why it happened with autotools and not with cmake.

Thanks!

I managed to figure our that's the actual problem is and how to bypass it, but I can't (and actually don't want to) dive too deep into the cause of it - it requires knowledge of the projects' source.

The problem isn't in any way related to the Autotools or even their usage in the project (which seems to be improper in some places). 

Real problem is that in https://github.com/audacity/audacity/blob/master/src/menus/ViewMenus.cpp `sizeof(wxFrame)` is 1016 bytes, whereas in https://github.com/audacity/audacity/blob/master/src/ProjectWindow.cpp same expression gives us only 1000 bytes. Wherefore all members of `ProjectWindow` class (which is indirect subclass of `wxFrame`) in `ViewMenus.cpp` got "moved" by 16 bytes and `GetVerticalScrollBar` at https://github.com/audacity/audacity/blob/master/src/menus/ViewMenus.cpp#L240 is returning reference into nowhere.

I don't know which size is the correct one; I don't know what causes such behaviour, but I suspect a dark magic with a memory that happens in some header files. Maybe there's something that adds two pointers for bookkeeping. Or maybe it's just messed up inclusion order of wxWidgets headers. I don't know.

To "fix" the problem you just need to include `ProjectWindow.h` before `Project.h` in https://github.com/audacity/audacity/blob/master/src/menus/ViewMenus.cpp#L10. That causes `sizeof(wxFrame)` in both files to align. The only question is, how many files there are where this happens too.

I've also checked on my machine; cmake build works either way (with the original code and wtih my "fix"), autotools build works only in the latter case. Probably it won't harm to make this "fix" permanent.

I don't know why Autotools build crashes, whereas cmake one doesn't. Probably there's something wrong with autotools config. I can see that at least some parts of your dark memory magic is conditional depending on C++ standard used, and cmake build uses c++14, while autotools - c++11. I've tried to change this and also added -NDEBUG to autotooled options. That made some interesting change: Audacity successfully performed track resize, but after that it crashed with the usual error dialog. Well, I don't want to investigate much further, 'cause I've found a workaround and that's good enough for me.

**Comment by SteveDaulton** [comment]
As we no longer use autotools, and you have a workaround that you say is satisfactory, I'll now close this issue.

**Comment by dvzrv** [comment]
@SteveDaulton ehm, well I for one do not have a satisfactory solution yet!

Can we at least link this issue to a proper pull request to *fix* this problem instead of just closing this and then forgetting about it?
If the "fix" is indeed so trivial, I can do a pull request later and test this. If there is no side-effect with the cmake setup I would really appreciate this being included!

@Barracuda72 thanks for looking into this more deeply!

**Comment by lllucius** [comment]
> 
> 
> > You mean that following good security practices to protect our sites from malicious actors goes against the tenants of free software? I'd absolutely love to read that article...please share a link.
> 
> Your mean, injecting proven to be malicious third-party service, compromising your website by your own hands and depriving users of their freedom is now called "good security practices"? WOW, that's really a long stretch...
> 
> If you want some links, you can just search for something like "Why Cloudflare is Bad". Here's only a couple:
> https://lists.gnu.org/archive/html/gnu-system-discuss/2017-03/msg00126.html
> http://techrights.org/2019/02/17/the-cloudflare-trap/
> https://pixie.garden/~/ThufiesBlog/don%27t-trust-cloud-flare (beware, certificate expired, but site itself isn't dubious and the article is very worth reading)
> 
Great articles!  But they still don't have anything to do with free software.

> > Prior to our removal of the autotools build, I'd tracked down the crash to somewhere between commits [a465ce0](https://github.com/audacity/audacity/commit/a465ce0046e1726b3d2d926091b91714e154a91e) (no crash) and [c1d129e](https://github.com/audacity/audacity/commit/c1d129ec9863a877cdd678695fff54a5dab5a191) (crash). I never did figure out why it happened with autotools and not with cmake.
> 
> Thanks!
> 
> I managed to figure our that's the actual problem is and how to bypass it, but I can't (and actually don't want to) dive too deep into the cause of it - it requires knowledge of the projects' source.
> 
> The problem isn't in any way related to the Autotools but probably to their usage in the project (which seems to be improper in some places).
> 
> Real problem is that in https://github.com/audacity/audacity/blob/master/src/menus/ViewMenus.cpp `sizeof(wxFrame)` is 1016 bytes, whereas in https://github.com/audacity/audacity/blob/master/src/ProjectWindow.cpp same expression gives us only 1000 bytes. Wherefore all members of `ProjectWindow` class (which is indirect subclass of `wxFrame`) in `ViewMenus.cpp` got "moved" by 16 bytes and `GetVerticalScrollBar` at https://github.com/audacity/audacity/blob/master/src/menus/ViewMenus.cpp#L240 is returning reference into nowhere.
> 
> I don't know which size is the correct one; I don't know what causes such behaviour, but I suspect a dark magic with a memory that happens in some header files. Maybe there's something that adds two pointers for bookkeeping. Or maybe it's just messed up inclusion order of wxWidgets headers. I don't know.
> 
Not THIS is something I can work with!  How on earth did you find it?!?!  I would NEVER have though to look for a size change.  I definitely look into it since the cause "may" still be lurking in there waiting to bite us in the future.

**Comment by SteveDaulton** [comment]
@dvzrv OK, I've reopened this issue, though personally I've lost track of what "this issue" is.

**Comment by JamesCrook** [comment]
@Barracuda72 Thank you for your work on this.  I have opened a separate issue for the Cloudflare points.

**Comment by dvzrv** [comment]
@SteveDaulton Thanks. I have just applied the fix suggested by @Barracuda72 and opened a pull request for this.

This little change indeed fixes the issue I am seeing and I will directly apply it also for the package on Arch Linux as well.

**Comment by lllucius** [comment]
This can also happen with the cmake build if you build with:

cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -Daudacity_use_pch=no ..

Still digging into the root cause...

**Comment by lllucius** [comment]
We have tracked down the root cause of the problem to this wxWidgets bug:

  https://trac.wxwidgets.org/ticket/15884
  https://github.com/wxWidgets/wxWidgets/commit/72aad0d6ea0c11c7cbb722eefb6106a7f1bdd203

I backported their fix to wxWidgets 3.0.5.1 so I could verify that it does indeed correct the issue:

  [weakref.patch.gz](https://github.com/audacity/audacity/files/4705113/weakref.patch.gz)

I've also included the patched weakref.h so it should be easy for you to test as well.
